### PR TITLE
FIX: Issue with params that have to ignored for DDP

### LINKED ISSUE #899: Modules_to_save can't be used with DDP and gradient checkpointing
### System Info

To train a model with gradient checkpointing and DDP, the model should have no unused parameters and the HF trainer's ddp_find_unused_parameters argument should be set to False.

However, peft's modules_to_save functionality creates unused parameters: the ModulesToSaveWrapper's original_module.

Gradient_checkpointing and DDP are really powerful in peft runs since large models can be made to fit on a single GPU, so it would be great to fix this functionality.

I attach a minimal example of a PEFT training script that fails with ddp.

### Who can help?

@pacman100 @younesbelkada @BenjaminBossan 

### Information

- [ ] The official example scripts
- [X] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder
- [ ] My own task or dataset (give details below)

### Reproduction

```python
from transformers import AutoTokenizer, AutoModelForCausalLM, Trainer
from peft import (
    TaskType,
    LoraConfig,
    get_peft_model,
    prepare_model_for_kbit_training
)
from transformers import DataCollatorForSeq2Seq, TrainingArguments
from datasets import load_dataset

model = AutoModelForCausalLM.from_pretrained('EleutherAI/pythia-70m',device_map='auto',torch_dtype='auto')
tokenizer = AutoTokenizer.from_pretrained('EleutherAI/pythia-70m')
tokenizer.pad_token_id = 0

model = prepare_model_for_kbit_training(model, use_gradient_checkpointing=True)

model = get_peft_model(model, LoraConfig(
    task_type=TaskType.CAUSAL_LM,
    inference_mode=False,
    r=8,
    lora_alpha=16,
    lora_dropout=0.1,
    bias="none",
    target_modules=['dense'],
    modules_to_save=['embed_out'],
)

dataset = load_dataset('tatsu-lab/alpaca', split="train[:100]")

train_data = dataset.map(lambda x: tokenizer(x['text'], return_tensors=None, padding='max_length', truncation=True, max_length=128))
train_data = train_data.map(lambda x: {'labels': x['input_ids']})

data_collator = DataCollatorForSeq2Seq(
    tokenizer=tokenizer,  pad_to_multiple_of=8, return_tensors="pt", padding=True
)

trainer = Trainer(
    model=model,
    train_dataset=train_data,
    args=TrainingArguments(output_dir='dummy', 	 
          per_device_train_batch_size=1,
          save_strategy='epoch',
          save_steps=1,
          ddp_find_unused_parameters=False),
    data_collator=data_collator)

trainer.train()
```

DDP launch command:
```bash
torchrun \
--nproc_per_node=1 \
--nnodes=1 \
--rdzv-backend=c10d \
--rdzv-endpoint=localhost:0 \
test.py
```

### Expected behavior

The training run should launch

### PR DESCRIPTION
**Pull request by BenjaminBossan**
Resolves #899

## Description

When using PyTorch `DistributedDataParallel`, there is an error if DDP wraps parameters that are not participating in the calculation of the loss. This is annoying, as we often have those types of parameters, e.g.:

- `ModulesToSaveWrapper` has a copy of the original weights
- When using multiple adapters, with only one being active

This PR aims at fixing this issue by exposing an attribute (actually: `property`) called `_ddp_params_and_buffers_to_ignore`. DDP uses this attribute to ignore the given parameters, making the error disappear.

## Implementation

Please let me know if there is a better way to implement this. The current solution has a few drawbacks:

Besides having to do this in the first place, it is also annoying that we have to implement this for all tuner layers, as they can differ when it comes to what parameters to ignore (we don't have a consistent naming scheme).

Also, our unit tests cannot cover this directly, as DDP requires a multi-gpu setup. Instead, the test is only for the existence, and correct value, of the `_ddp_params_and_buffers_to_ignore` attribute. Writing a proper DDP test would require more effort but should be feasible.

We also rely on a private attribute here, so there is a danger of this breaking with future PyTorch releases.

## DONE/TODO:

- [x] ModulesToSaveWrapper
- [x] LoRA
- [ ] IA³
- [ ] AdaLora

### CODE REVIEW

File: src/peft/peft_model.py [hunk 1]
```diff
@@ -125,6 +125,23 @@ def __init__(self, model: PreTrainedModel, peft_config: PeftConfig, adapter_name
...
+        # Add an attribute _ddp_params_and_buffers_to_ignore. This is sometimes required for DDP, as we may
+        # want to ignore certain parameters to avoid the error "Parameters which did not receive grad for rank X".
+        # See issue 899.
+        ddp_params_and_buffers_to_ignore = set()
+        for name, module in self.named_modules():
+            if module is self:
+                # avoid infinite recursion
+                continue
+
+            module_params_and_buffers_to_ignore = getattr(module, "_ddp_params_and_buffers_to_ignore", [])
```
Comment (younesbelkada) [inline src/peft/peft_model.py:139]:
```suggestion
            if isinstance(module, BaseTunerLayer):
                module_params_and_buffers_to_ignore = module._ddp_params_and_buffers_to_ignore
```

Then inside the if block perform the for loop

**Review by younesbelkada — COMMENTED** [review]
Thanks a lot @BenjaminBossan for this very elegant way of solving the issue you have stated.
IMO we can push that further and make it more future proof if we don't use getattr.  I think `getattr` with `[]` passed as the last argument can silently lead to always getting empty lists; what about adding an abstract property inside `BaseTunerLayer`, so that we could safely just use `if isinstance(module, BaseTunerLayer)` as a condition to retrieve the params and bufffers to ignore of that module. What do you think?

**Comment by BenjaminBossan** [comment]
This PR is superseded by #905, I forgot to close it.
